File: ListNode/2addTwoNumbers.py
#ListNode

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        l3 = temp = ListNode(0)
        carry = 0
        
        while l1 or l2 or carry:
            n1, n2 = 0, 0
            if l1:
                n1 = l1.val
                l1 = l1.next
            if l2:
                n2 = l2.val
                l2 = l2.next
            s = n1 + n2 + carry
            carry = s/10
            remainder = s%10
            temp.next = ListNode(remainder)
            temp = temp.next
        return l3.next

File: LinkedList/2AddTwoNumbers.cpp
/**
 Solution:
 Use recursion to pass every node of l1 and l2. If the sum is equal or greater than 10, pass 1 and rev->next to function again.
 Base case: (!l1 && !l2), (!l1), (!l2).
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        if(!l1 && !l2) return NULL;
        if(!l1) return l2;
        if(!l2) return l1;
        int sum = l1->val + l2->val;
        ListNode *rev = new ListNode(sum%10);
        rev->next = addTwoNumbers(l1->next, l2->next);
        if(sum>=10) rev->next = addTwoNumbers(rev->next, new ListNode(1));
        return rev;
    }
};